toml/decoder.py

```diff
@@ -957,3 +1021,114 @@ def load_array(self, a):
         return retval
+
+    def handle_comment(self, currentlevel, line, comment_type,
+                       original, recent=None):
+        pass
+
+    def store_comment_info(self, i, comment_info, comment_lines, original):
+        pass
+
+
+class TomlPreserveCommentDecoder(TomlDecoder):
+
+    def __init__(self, _dict=dict):
+        super(TomlPreserveCommentDecoder, self).__init__(_dict)
+
+    def store_comment_info(self, i, comment_info, comment_lines, original):
+        # Store the line number of comments in the original line
+        # along with the number of lines the comment spans into
+
+        line_no = original.count('\n', 0, i) + 1
+        ls = original.split('\n')
+        comment_line = ls[line_no - 1]
+
+        if comment_line.strip().startswith('#'):
+            comment_info.append([line_no, 0])
+            comment_lines.append(line_no)
+            return
+
+        pos = comment_line.find('#')
+        str_lst = list(comment_line)
+        while pos < len(str_lst):
+            str_lst[pos] = ''
+            pos += 1
+        comment_line = ''.join(str_lst)
+        if ('=' in comment_line) and (len(comment_line.split('=')) == 2) \
+                and (not comment_line.split('=')[1].strip().startswith('[')):
+            comment_info.append([line_no, 0])
+            comment_lines.append(line_no)
+        else:
+            current_line_no = line_no - 1
+            while (len(comment_line.split('=')) != 2) or \
+                    (not comment_line.split('=')[1].strip().startswith('[')):
+                current_line_no = current_line_no - 1
+                comment_line = ls[current_line_no]
+            start = current_line_no + 1
+            array_len = 0
+            while not comment_line.strip().endswith(']'):
+                array_len += 1
+                current_line_no += 1
+                comment_line = ls[current_line_no]
+            if [start, array_len + 1] not in comment_info:
+                comment_info.append([start, array_len + 1])
+                comment_lines.append(line_no)
+
+    @staticmethod
+    def handle_inline_comment(line, currentlevel):
+        pos = line.find('#')
+        k = pos
+        comment = ""
+        try:
+            while line[k] != '\n':
+                comment += line[k]
+                k += 1
+        except IndexError:
+            pass
+        pair = line.split('=')
+        currentlevel[CommentKey(pair[0].strip() + '-' +
+                                'comment' + '_inline', comment)] = comment
+
+    def handle_comment(self, currentlevel, line, comment_type,
+                       original, recent=None):
+        if comment_type == 'normal':
+            currentlevel[CommentKey(recent + '-' + 'comment', line)] = line
+        elif comment_type == 'inline':
+            self.handle_inline_comment(line, currentlevel)
+        else:
+            self.handle_array_comment(line, currentlevel, original)
+
+    @staticmethod
```

Comment:
I'm not sure I understand this decorator.
It says static method, but you are calling this as if it is an instance method.
Would it not be easier to understand if you called it as `handle_array_comment` (not `self.handle_array_comment`) and dropped this decorator?